cpp/src/copying/contiguous_split.cu

```diff
@@ -14,465 +14,923 @@
+  {
+    return 0;
+  }
+
+  template <typename T>
+  constexpr std::enable_if_t<is_fixed_width<T>(), int> __device__ operator()() const noexcept
+  {
+    return sizeof(T);
+  }
+};
 
 };  // anonymous namespace
 
+namespace detail {
+
 std::vector<contiguous_split_result> contiguous_split(cudf::table_view const& input,
                                                       std::vector<size_type> const& splits,
                                                       rmm::mr::device_memory_resource* mr,
                                                       cudaStream_t stream)
 {
-  auto subtables = cudf::split(input, splits);
+  if (input.num_columns() == 0) { return {}; }
+  if (splits.size() > 0) {
+    CUDF_EXPECTS(splits.back() <= input.column(0).size(),
+                 "splits can't exceed size of input columns");
+  }
+  {
+    size_type begin = 0;
+    for (size_t i = 0; i < splits.size(); i++) {
+      size_type end = splits[i];
+      CUDF_EXPECTS(begin >= 0, "Starting index cannot be negative.");
+      CUDF_EXPECTS(end >= begin, "End index cannot be smaller than the starting index.");
+      CUDF_EXPECTS(end <= input.column(0).size(), "Slice range out of bounds.");
+      begin = end;
+    }
+  }
 
-  // optimization : for large numbers of splits this allocation can dominate total time
-  //                spent if done inside alloc_and_copy().  so we'll allocate it once
-  //                and reuse it.
-  //
-  //                benchmark:        1 GB data, 10 columns, 256 splits.
-  //                no optimization:  106 ms (8 GB/s)
-  //                optimization:     20 ms (48 GB/s)
-  rmm::device_vector<column_split_info> device_split_info(input.num_columns());
+  size_t const num_partitions = splits.size() + 1;
 
-  std::vector<contiguous_split_result> result;
-  std::transform(subtables.begin(),
-                 subtables.end(),
-                 std::back_inserter(result),
-                 [mr, stream, &device_split_info](table_view const& t) {
-                   return alloc_and_copy(t, device_split_info, mr, stream);
+  // if inputs are empty, just return num_partitions empty tables
+  if (input.column(0).size() == 0) {
+    std::vector<contiguous_split_result> result;
+    result.reserve(num_partitions);
+
+    auto iter = thrust::make_counting_iterator(0);
+    std::transform(
+      iter, iter + num_partitions, std::back_inserter(result), [&input](int partition_index) {
+        return contiguous_split_result{input, std::make_unique<rmm::device_buffer>()};
+      });
+
+    return std::move(result);
+  }
+
+  size_t const num_root_columns = input.num_columns();
+
+  // compute # of source buffers (column data, validity, children), # of partitions
+  // and total # of buffers
+  size_type const num_src_bufs = count_src_bufs(input.begin(), input.end());
+  size_t const num_bufs        = num_src_bufs * num_partitions;
+
+  // clang-format off
```

Comment:
I don't think this qualifies as code where it's objectively better to turn formatting off...

```suggestion
```